### DISCUSSION TITLE
spawning new wezterm pane via unix window manager hotkeys

### DISCUSSION THREAD

**Discussion by dch** [wezterm/wezterm#1486] [Q&A] [answered]
I'm switching from urxvt to wezterm, so far all good but
I can't figure out how to use an i3 hotkey to spawn a new
wezterm, in its own pane.

So far I'm getting started as follows:

```
$ wezterm-mux-server --daemonize
$ wezterm-gui start &
$ wezterm cli list
WINID TABID PANEID SIZE   TITLE              CWD                                         
    0     0      0 158x79 fish /usr/home/dch file://wintermute/usr/home/dch
```

Inside this existing wezterm, `{key="n", mods="CTRL|SHIFT", action="SpawnWindow"},`
works well, and running `wezterm cli --spawn --pane-id 0` or
`wezterm cli --spawn --window-id 0` does what I'd expect.

From a non-wezterm, running:

```
$ env WEZTERM_UNIX_SOCKET=/tmp/dch.abc123/wezterm/sock wezterm cli spawn --pane-id 0
```

works, but this requires getting the socket first, which is
ok but messy.

Is there a better way to do this, from outside the wezterm
env in the window manager?

**Comment by wez** [comment, marked as answer]
I'd suggest using this in your config (taken from https://wezfurlong.org/wezterm/multiplexing.html)

```lua
return {
  unix_domains = {
    {
      name = "unix",
    }
  },

  -- This causes `wezterm` to act as though it was started as
  -- `wezterm connect unix` by default, connecting to the unix
  -- domain on startup.
  default_gui_startup_args = {"connect", "unix"},
}
```

This will result in everything using the same stable unix domain socket and will make `wezterm-gui` connect to that instance.

**Reply by dch** [reply]
thanks! I was running an earlier version of wezterm, and it doesn't
support this. After upgrading, have I misunderstood how this is
supposed to work entirely? It seems to work just fine, without
any `wezterm-mux-server`.

- for each new X11 wezterm "window" I am starting `wezterm-gui start`

which gives me for 4 wezterms:

```
$ l (find /tmp/$USER.*/wezterm)
srwx------  1 dch  wheel     0B Jan  3 12:33 /tmp/dch.ulClS2F2/wezterm/gui-sock-73768=
srwx------  1 dch  wheel     0B Jan  3 12:36 /tmp/dch.ulClS2F2/wezterm/gui-sock-77414=
srwx------  1 dch  wheel     0B Jan  3 12:33 /tmp/dch.ulClS2F2/wezterm/gui-sock-84334=
srwx------  1 dch  wheel     0B Jan  3 12:33 /tmp/dch.ulClS2F2/wezterm/gui-sock-95103=
```

do I even need to start an initial `wezterm-mux-server` at all?

**Reply by wez** [reply]
I've made some changes in `main` today that mean that the normal gui operation (without configuring `unix_domains`) is now discoverable to `wezterm cli` without you having to figure out the socket path.

So `wezterm cli spawn --new-window` should work without any extra effort.
The same mechanism is used to make `wezterm` (and the underlying `wezterm-gui` process) open new windows into an already running instance of wezterm.

For spawning a tab into the active window: some work needs to be done in the mux protocol/data model to allow it to know which pane/tab/window has focus in the GUI so that we can pick a good default for `--pane-id`.

**Reply by dch** [reply]
thats fantastic. many thanks!

**Reply by dch** [reply]
to clarify, what's the benefit of starting with `wezterm-mux-server` at all? it seems `wezterm-gui start` alone is sufficient?

**Reply by wez** [reply]
If you only ever work from the desktop environment of the local machine then you don't need to run a separate mux server, and you'll have a slightly better experience.

The mux server is helpful if you want to connect multiple machines to the same server, or if you want to have your wezterm workspace keep running with the same set of windows and tabs even if you log out of the desktop environment and eg: start a different window manager.
